File: python/converter.py
# To run: python converter.py ../input.md
# Pros:
#   - Can run/test program with one command
#   - Really good file IO and string support
#   - Modular indexing is helpful for getting the last element, no bounds checks required, may be confusing if unintentional though
#   - Being able to say 'elem in set' as a boolean is useful
#   - Sets in general have really good support, very easy, no cognition required
#   - Formatted strings!
# Cons:
#   - No implicit main function, have to do that stupid 'if __name__ == __main__' thing
#   - Can't leave a function body/if body/for body blank when filling out a function, which is sorta annoying
#   - Command line arguments are weird, because there's no implicit main function
#   - Inheritance is a little confusing, I prefer interfaces instead
#   - Dynamically typed
#   - Still have to order class definitions (may be a lang server thing)
#   - Indentation makes it hard to read sometimes, hard to hold onto 'landmarks'


import sys


class MarkdownNode:
    pass

class Header(MarkdownNode):
    def __init__(self, size, children):
        self.size = size
        self.children = children
    
    def __str__(self):
        text = f'<h{self.size}>'
        for i, child in enumerate(self.children):
            if i == 0:
                text += str(child).lstrip()
            else:
                text += str(child)
        return text + f'</h{self.size}>\n'

class Paragraph(MarkdownNode):
    def __init__(self, children):
        self.children = children
    
    def __str__(self):
        text = '<p>'
        for child in self.children:
            text += str(child)
        return text + '</p>\n\n'

class CodeBlock(MarkdownNode):
    def __init__(self, text):
        self.text = text
    
    def __str__(self):
        return f'<pre><code>{self.text}</code></pre>\n\n'

class Image(MarkdownNode):
    def __init__(self, text, url):
        self.text = text
        self.url = url
    
    def __str__(self):
        return f'<img src=\'{self.url}\' alt=\'{self.text}\' />\n\n'

class Text(MarkdownNode):
    def __init__(self, text):
        self.text = text
    
    def __str__(self):
        return self.text

class Bold(MarkdownNode):
    def __init__(self, children):
        self.children = children
    
    def __str__(self):
        text = '<strong>'
        for child in self.children:
            text += str(child)
        return text + '</strong>'

class Italic(MarkdownNode):
    def __init__(self, children):
        self.children = children
    
    def __str__(self):
        text = '<em>'
        for child in self.children:
            text += str(child)
        return text + '</em>'

class Code(MarkdownNode):
    def __init__(self, text):
        self.text = text
    
    def __str__(self):
        return f'<code>{self.text}</code>'

class Link(MarkdownNode):
    def __init__(self, text, url):
        self.text = text
        self.url = url
    
    def __str__(self):
        return f'<a href=\'{self.url}\'>{self.text}</a>'


class Token:
    def __init__(self, data, line, col):
        self.data = data
        self.line = line
        self.col = col

tokens = []
index = 0

def init_parser(contents:str):
    data = contents[0]
    line = 1
    col = 1
    old_col = 1
    old_c = data
    for c in contents[1:]:
        if old_c in '\n\r' or ((old_c in '_*`#[]()!') != (c in '_*`#[]()!')) or c in '#[(!\n\r':
            tokens.append(Token(data, line, old_col))
            old_col = col + 1
            if c == '\n':
                line += 1
                col = 0
            data = ''
        if old_c != '#' or not c.isspace():
            data += c
        col += 1
        old_c = c
    tokens.append(Token(data, line, col))
    tokens.append(Token('\n', line, col))


def pop():
    global index
    index += 1
    return tokens[index - 1]

def peek():
    return tokens[index]
    
def accept(data):
    if tokens[index].data == data:
        return pop().data
    else:
        return None

def expect(data):
    if accept(data) == None:
        top = peek()
        print(f'error: {top.line}:{top.col} expected `{data}` got `{top.data if top.data[0] in "_*`#[]()!" else "text"}`')
        sys.exit(1)

def take_until(sentinel):
    text = ''
    while not accept(sentinel):
        text += pop().data
    return text

def parse_document():
    while index < len(tokens) - 1:
        node = parse_node()
        if node:
            yield node

def parse_node():
    if accept('#'):
        return parse_header()
    elif accept('```'):
        return parse_code_block()
    elif accept('!'):
        return parse_image()
    elif accept('\n'):
        return None
    else:
        return parse_paragraph()

def parse_header():
    size = 1
    while accept('#'):
        size += 1
    return Header(size, parse_formatted_text({'\n'}))

def parse_paragraph():
    return Paragraph(parse_formatted_text({'\n'}))

def parse_code_block():
    return Code(take_until('```'))

def parse_image():
    expect('[')
    text = pop().data
    expect(']')
    expect('(')
    url = pop().data
    expect(')')
    return Image(text, url)

def parse_formatted_text(bounds):
    retval = []
    while not peek().data in bounds:
        if accept('_') or accept('*'):
            retval.append(parse_italic(bounds))
        elif accept('__') or accept('**'):
            retval.append(parse_bold(bounds))
        elif accept('`'):
            retval.append(parse_code())
        elif accept('['):
            retval.append(parse_link())
        else:
            retval.append(Text(pop().data))
    return retval

def parse_italic(bounds):
    children = parse_formatted_text(bounds.union({'*', '_'}))
    if not accept('*'):
        expect('_')
    return Italic(children)

def parse_bold(bounds):
    children = parse_formatted_text(bounds.union({'**', '__'}))
    if not accept('**'):
        expect('__')
    return Bold(children)

def parse_code():
    return Code(take_until('`'))

def parse_link():
    text = pop().data
    expect(']')
    expect('(')
    url = pop().data
    expect(')')
    return Link(text, url)


# Open input file, read contents
with open(sys.argv[1]) as f:
    contents = f.read()

# Create parser, parse Markdown document
init_parser(contents)

# Open output file, write HTML document
with open('output.html', 'w') as f:
    for node in parse_document():
        f.write(str(node))

File: javascript/converter.js
// To run: node converter.js ../input.md
// Pros: 
//  - No setup required, just a .js file
//  - Don't always need semicolons
//  - Nice way to import stuff
//  - Optionals (for typescript)
//  - JSON for every object, can print out an object and have it be formatted pretty well, great for debugging
//  - Implicit undefined return along with being able to check for undefined as a boolean is concise
//  - Prototypical classes are basically just HashMaps that can contain functions. HashMaps + first class functions + syntactic sugar = prototypical classes
//  - Lambdas and closures
//  - Spread operator
// Cons:
//  - Need parens around if conditions
//  - Couldn't get TypeScript working because of dependency issues (often the case!)
//  - Undefined values if values aren't passed to a function. No warning, no error. (for JavaScript)
//  - Dynamically typed
//  - explicit `this`
//  - weird distinction between `of` and `in` for foreach loops
//  - Weird triple eq thing. Makes sense for web development i suppose. == and === should have been flipped IMO
//  - No static checking unfortunately

fs = require('fs')

class Header {
    constructor (size, children) {
        this.size = size
        this.children = children
    }

    getHTML() {
        let text = '<h' + this.size + '>'
        for (let child of this.children) {
            text += child.getHTML()
        }
        return text + '</h' + this.size + '>\n'
    }
}

class Paragraph {
    constructor (children) {
        this.children = children
    }

    getHTML() {
        let text = '<p>'
        for (let child of this.children) {
            text += child.getHTML()
        }
        return text + '</p>\n\n'
    }
}

class CodeBlock {
    constructor (text) {
        this.text = text
    }

    getHTML() {
        return '<pre><code>' + this.text + '</code></pre>\n\n'
    }
}

class Image {
    constructor (text, url) {
        this.text = text
        this.url = url
    }

    getHTML() {
        return '<img src="' + this.url + '" alt="' + this.text + '" />\n'
    }
}

class Text {
    constructor (text) {
        this.text = text
    }

    getHTML() {
        return this.text
    }
}

class Italic {
    constructor (children) {
        this.children = children
    }

    getHTML() {
        let text = '<em>'
        for (let child of this.children) {
            text += child.getHTML()
        }
        return text + '</em>'
    }
}

class Bold {
    constructor (children) {
        this.children = children
    }

    getHTML() {
        let text = '<strong>'
        for (let child of this.children) {
            text += child.getHTML()
        }
        return text + '</strong>'
    }
}

class Code {
    constructor (text) {
        this.text = text
    }

    getHTML() {
        return '<code>' + this.text + '</code>'
    }
}

class Link {
    constructor (text, url) {
        this.text = text
        this.url = url
    }

    getHTML() {
        return '<a href="' + this.url + '">' + this.text + '</a>'
    }
}

class Token {
    constructor (data, line, col) {
        this.data = data
        this.line = line
        this.col = col
    }
}

function isSpecialChar(c) {
    return c == '_' ||
        c == '*' ||
        c == '`' ||
        c == '#' ||
        c == '[' ||
        c == ']' ||
        c == '(' ||
        c == ')' ||
        c == '!'
}

class Parser {
    constructor (contents) {
        this.tokens = []
        this.index = 0

        let data = contents[0]
        let line = 1
        let col = 1
        let oldCol = 1
        let oldC = contents[0]
        for (let c of contents.substring(1)) {
            if (oldC == '\n' || oldC == '\r' || (isSpecialChar(oldC) != isSpecialChar(c)) || c == '#' || c == '[' || c == '(' || c == '!' || c == '\n' || c == '\r') {
                if (!data.includes('\r')) {
                    this.tokens.push(new Token(data, line, oldCol))
                }
                oldCol = col + 1
                if (c == '\n') {
                    line ++
                    col = 0
                }
                data = ''
            }
            if (oldC != '#' || !(/\s/g.test(c))) {
                data += c
            }
            col++
            oldC = c
        }
        this.tokens.push(new Token(data, line, col))
        this.tokens.push(new Token('\n', line, col))
    }

    pop() {
        this.index++
        return this.tokens[this.index - 1]
    }

    peek() {
        return this.tokens[this.index]
    }

    accept(data) {
        if (this.peek().data === data) {
            return this.pop()
        }
    }

    expect(data) {
        if (!this.accept(data)) {
            let top = this.peek()
            if (isSpecialChar(top.data[0])) {
                console.log("error: %d:%d expected `%s`, got `%s`\n", top.line, top.col, token, top.data)
            } else {
                console.log("error: %d:%d expected `%s`, got text\n", top.line, top.col, token)
            }
            process.exit(1)
        }
    }

    parseDocument() {
        let retval = []
        while (this.index < this.tokens.length - 1) {
            let node = this.parseNode()
            if (node) {
                retval.push(node)
            }
        }
        return retval
    }

    parseNode() {
        if (this.accept('#')) {
            return this.parseHeader()
        } else if (this.accept('```')) {
            return this.parseCodeBlock()
        } else if (this.accept('!')) {
            return this.parseImage()
        } else if (!this.accept('\n')) {
            return this.parseParagraph()
        }
    }

    parseHeader() {
        let size = 1
        while (this.accept('#')) {
            size++
        }
        return new Header(size, this.parseFormattedText(['\n']))
    }

    parseParagraph() {
        return new Paragraph(this.parseFormattedText(['\n']))
    }

    parseCodeBlock() {
        let text = ''
        while (!this.accept('```')) {
            text += this.pop().data
        }
        return new CodeBlock(text)
    }

    parseImage() {
        this.expect('[')
        let text = this.pop().data
        this.expect(']')
        this.expect('(')
        let url = this.pop().data
        this.expect(')')
        return new Image(text, url)
    }

    parseFormattedText(bounds) {
        let retval = []
        while (!bounds.find(item => { return item === this.peek().data } )) {
            if (this.accept('_') || this.accept('*')) {
                retval.push(this.parseItalic(bounds))
            } else if (this.accept('__') || this.accept('**')) {
                retval.push(this.parseBold(bounds))
            } else if (this.accept('`')) {
                retval.push(this.parseCode(bounds))
            } else if (this.accept('[')) {
                retval.push(this.parseLink(bounds))
            } else {
                retval.push(new Text(this.pop().data))
            }
        }
        return retval
    }

    parseItalic(bounds) {
        let children = this.parseFormattedText([...bounds, '*', '_'])
        if (!this.accept('*')) {
            this.expect('_')
        }
        return new Italic(children)
    }

    parseBold(bounds) {
        let children = this.parseFormattedText([...bounds, '**', '__'])
        if (!this.accept('**')) {
            this.expect('__')
        }
        return new Bold(children)
    }

    parseCode() {
        let text = ''
        while (!this.accept('`')) {
            text += this.pop().data
        }
        return new Code(text)
    }

    parseLink() {
        let text = this.pop().data
        this.expect(']')
        this.expect('(')
        let url = this.pop().data
        this.expect(')')
        return new Link(text, url)
    }
}

if (process.argv.length != 3) {
    console.log('usage: node converter.js <markdown-filename>')
} else {
    // Open input file, read contents
    try {  
        var contents = fs.readFileSync(process.argv[2], 'utf8'); 
    } catch(e) {
        console.log('Error:', e.stack);
        process.exit(1)
    }

    // Create parser, parse Markdown document
    let parser = new Parser(contents)
    let document = parser.parseDocument()

    // Open output file, write HTML document
    let output = ''
    for (let node of document) {
        output += node.getHTML()
    }
    try {
        fs.writeFileSync('output.html', output)
    } catch (err) {
        console.error(err);
        process.exit(1);
    }
}